### PR TITLE
Add summary table and charts to the Dashboard

### PR DESCRIPTION
**Pull request by cuducos**
**What is the purpose of this Pull Request?**
Based on @hakib's post (many thanks!) [How to turn Django Admin into a lightweight dashboard](https://hakibenita.com/how-to-turn-django-admin-into-a-lightweight-dashboard) I suggested an implementation for Jarbas. This might be useful for non-tech people interested in an simple overview of CEAP (cc @tatianasb and @vilapedro).

**What was done to achieve this purpose?**
Basically followed the steps in @hakib's post.

**How to test if it really works?**
Run Jarbas, reach `/dashboard/chamber_of_deputies/reimbursementsummary/` and expect I expect you can browse in an interface like this one:
image

**Who can help reviewing it?**
Just thinking out loud depending on each one's interest and availability: @jtemporal @giovanisleite @cabral @turicas @irio @leportella

### CODE REVIEW

**Comment by willianpaixao** [comment]
I can only review the Python and Jinja files. It looks good to me.

**Comment by cuducos** [comment]
> @cuducos do you know if it's possible to write tests to such feature?

Yes. If I'm not wrong [this talk](https://youtu.be/41ek3VNx_6Q) gives insightful ideas about how to test Admin views. However Jarbas Django Admin parts are largely untested here. Unfortunately.

File: jarbas/dashboard/admin/__init__.py [hunk 1]
```diff
@@ -149,4 +155,80 @@ def get_search_results(self, request, queryset, search_term):
...
+            size = (total - low) / (high - low) * 100
+
+        corrected_size = Decimal('5') + (Decimal('0.95') * size)
+        return corrected_size
+
+    def changelist_view(self, request, extra=None):
+        response = super().changelist_view(request, extra_context=extra)
+
+        try:
+            queryset = response.context_data['cl'].queryset
```
Comment (hakib) [inline jarbas/dashboard/admin/__init__.py]: Since I wrote the article I found that Django will execute two more unnecessary queries on the underlying model (in your case it's `Reimbursement`). If the model holds a lot of data, these queries can be eliminated, but it requires some dirty hacks. 

If you find performance to be an issue in this specific page, you can eliminate the two extra queries.

The two queries original from [within the Django admin](https://github.com/django/django/blob/master/django/contrib/admin/options.py#L1780). To eliminate them you need to fool Django into thinking the queryset is empty. But, you still want Django to pass the queryset through the `ChangeList` and apply all the filters (from `date_hierarchy` and `list_filters`).

First, you need to override the `get_queryset` function of the `ReimbursementSummaryModelAdmin` class:

```python
def get_queryset(self, request):
        return super().get_queryset(request)
```

To fool Django into thinking the queryset is empty,  you need to return a queryset that will return no results, but one that you can reverse later (after the filters were applied):

```python
def add_false_condition(qs):
    return qs.extra(where=['1=0'])

...

class ReimbursementSummaryModelAdmin(PublicAdminModelAdmin): 
   ...
   def get_queryset(self, request):
        return add_false_condition(super().get_queryset(request).using(settings.READ_REPLICA_USING))
```

This will add the false condition "1=0" to the query. Some vendors, such as PostgreSQL will not do anything when they see such condition. This is good, this is what we want.

Now, you still need the queryset with all it's filters when you fetch the data for the list view. This is where we remove the fals condition:

```python
def remove_false_condition(qs):
    clone = qs.all()
    clone.query.where.children = [
        e for e in qs.query.where.children
        if not hasattr(e, 'sqls') or e.sqls != ['1=0']
    ]
    return clone

...

class ReimbursementSummaryModelAdmin(PublicAdminModelAdmin): 
   ...
   def changelist_view(self, request, extra_context=None):
       response = super().changelist_view(request, extra_context=extra_context)

      try:
          qs = response.context_data['cl'].queryset
      except (AttributeError, KeyError):
          return response

       qs = remove_false_condition(qs)
```

Feel free to use this but be aware that this is a really dirty hack. I only had to use this hack after our app got *really* big.

File: jarbas/dashboard/templatetags/dashboard.py [hunk 2]
```diff
@@ -11,3 +13,47 @@ def rename_title(title):
...
+def brazilian_reais(value):
+    return f'R$ {brazilian_float(value)}'
+
+
+@register.filter()
+def brazilian_float(value):
+    # TODO properly use locales
+    value = value or 0
+    value = f'{value:,.2f}'
+    return value.replace(',', 'x').replace('.', ',').replace('x', '.')
```
Comment (hakib) [inline jarbas/dashboard/templatetags/dashboard.py]: Found this function interesting so thought I see what it does. 

Another option is using `translate`:

```
>>> brazilian_float_trans = str.maketrans(',.', '.,')                      

>>> '{:,.2f}'.format(10000 / 3).translate(brazilian_float_trans)           
'3.333,33'

>>> '{:,.2f}'.format(10000 / 3)                                            
'3,333.33'
```

File: jarbas/dashboard/admin/__init__.py [hunk 3]
```diff
@@ -149,4 +155,80 @@ def get_search_results(self, request, queryset, search_term):
...
+            .values('period')
+            .annotate(total=Sum('total_net_value'))
+            .order_by('period')
+        )
+        summary_range = summary_over_time.aggregate(
+            low=Min('total'),
+            high=Max('total')
+        )
+        high = summary_range.get('high', 0)
+        low = summary_range.get('low', 0)
```
[the remark this answers is not in the document]
Comment (cuducos) [reply jarbas/dashboard/admin/__init__.py]: Hum… interesting. After all scaling the application is easier (and cheaper) than scaling the database 👍  Done in ab46a26

File: jarbas/dashboard/admin/__init__.py (hunk 1 above)
Comment (cuducos) [reply jarbas/dashboard/admin/__init__.py]: We do have a bottleneck in Postgres, so we use a some caching strategies. For this summary I'm caching the whole context we calculate in b0c4ab3. That should be enough, but gonna monitor performance once it rolls out.

File: jarbas/dashboard/templatetags/dashboard.py (hunk 2 above)
Comment (cuducos) [reply jarbas/dashboard/templatetags/dashboard.py]: Wow! That's way better! Refactored in 0eb34fe

**Comment by cuducos** [comment]
Many thanks @willianpaixao and @hakib for the feedback. I believe addressed most of the issues and now we have a better _pull request_, yay ❤️

I did a quite relevant change dismissing `data_hierarchy`. As explained in 7457961 :

> The `date_hirearchy` method leads to doubts as `issue_date` is highly imprecise (there are expenses from 1959 entil 2050, or something like that). Also this might lead to confusion when reading the summary.

Now the UI looks like that and users can browser using regular list filters from the standard dashboard:

image

**Comment by giovanisleite** [comment]
Whooa, this is so great 🎆

File: jarbas/dashboard/admin/__init__.py [hunk 4]
```diff
@@ -149,4 +157,120 @@ def get_search_results(self, request, queryset, search_term):
...
+        list_filters.SuspiciousListFilter,
+        list_filters.HasReimbursementNumberFilter,
+        list_filters.StateListFilter,
+        list_filters.YearListFilter,
+        list_filters.MonthListFilter,
+        list_filters.DocumentTypeListFilter,
+    )
+
+    def get_period(self, request):
+        if 'year' in request.GET:
```
Comment (giovanisleite) [inline jarbas/dashboard/admin/__init__.py]: This is confuse but I don't have a better suggestion 🦀
Maybe replace the `get_period` by something less context dependent. What do you think?

Comment (cuducos) [reply jarbas/dashboard/admin/__init__.py]: Actually I think I did the opposite: I changed the name for something more context specific. But I added nice docstrings to help the reader… what do you think? It's basically c1ae375

Comment (giovanisleite) [reply jarbas/dashboard/admin/__init__.py]: This is a better idea, very nice, @cuducos 🚀
